go/cmd/ensureconda.go

```diff
@@ -0,0 +1,229 @@
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"github.com/Wessie/appdirs"
+	"github.com/hashicorp/go-version"
+	"github.com/spf13/cobra"
+	"os"
+	"runtime"
+	"strconv"
+)
+
+var (
+	// Used for flags.
+
+	rootCmd = &cobra.Command{
+		Use:   "ensureconda",
+		Short: "",
+		Long:  ``,
+		Run: func(cmd *cobra.Command, args []string) {
+			mamba, err := evaluateFlagPair(cmd, "mamba")
+			if err != nil {
+				panic(err)
+			}
+			micromamba, err := evaluateFlagPair(cmd, "micromamba")
+			if err != nil {
+				panic(err)
+			}
+			conda, err := evaluateFlagPair(cmd, "conda")
+			if err != nil {
+				panic(err)
+			}
+			condaExe, err := evaluateFlagPair(cmd, "conda-exe")
+			if err != nil {
+				panic(err)
+			}
+			noInstall, err := cmd.Flags().GetBool("no-install")
+			if err != nil {
+				panic(err)
+			}
+
+			executable, err := EnsureConda(mamba, micromamba, conda, condaExe, true)
+			if executable != "" {
+				fmt.Print(executable)
+				os.Exit(0)
+			}
+			if !noInstall {
+				executable, err = EnsureConda(mamba, micromamba, conda, condaExe, noInstall)
+				if err != nil {
+					er(err)
+				}
+				if executable != "" {
+					fmt.Print(executable)
+					os.Exit(0)
+				}
+			}
+			os.Exit(1)
+		},
+	}
+)
+
+var TestSitePath string
+
+func sitePath() string {
+	if TestSitePath != "" {
+		return TestSitePath
+	}
+	return appdirs.UserDataDir("ensure-conda", "", "", false)
+}
+
+func EnsureConda(mamba bool, micromamba bool, conda bool, condaExe bool, noInstall bool) (string, error) {
+	var executable string
+	dataDir := sitePath()
+	minMambaVersion, _ := version.NewVersion("0.7.3")
+	minCondaVersion, _ := version.NewVersion("4.8.2")
+
+	mambaVersionCheck := executableHasMinVersion(minMambaVersion, "mamba")
+	microMambaVersionCheck := executableHasMinVersion(minMambaVersion, "")
+	condaVersionCheck := executableHasMinVersion(minCondaVersion, "conda")
+
+	if mamba {
+		executable, _ = ResolveExecutable("mamba", dataDir, mambaVersionCheck)
+		if executable != "" {
+			return executable, nil
+		}
+	}
+	if micromamba {
+		executable, _ = ResolveExecutable("micromamba", dataDir, microMambaVersionCheck)
+		if executable != "" {
+			return executable, nil
+		}
+		if !noInstall {
+			exe, err := InstallMicromamba()
+			if err != nil {
+				return "", err
+			}
+			if valid, _ := microMambaVersionCheck(exe); valid {
+				return exe, nil
+			}
+		}
```

Comment:
hmm, we could use an interface (Install, versionCheck, ResolveExecutable) for each micromamba, condaExe, etc.